### PR TITLE
add missing semigroup and monoid instances

### PR DESCRIPTION
**Pull request by bassjacob**
putting this up as a WIP. tests to come

### CODE REVIEW

**Comment by bassjacob** [comment]
Oh sorry, I knew I was going on vacation and didn't want to leave it open while I was working on it :smile: still going through with it, should get some good time into it this week.

**Comment by bassjacob** [comment]
well, that was a total lie. I should be good to close this off over the next few days if it's still needed

File: modules/core/arrow-data/src/test/kotlin/arrow/data/Function0Test.kt [hunk 1]
```diff
@@ -19,9 +24,19 @@ class Function0Test : UnitSpec() {
   init {
     ForFunction0 extensions {
       testLaws(
+        SemigroupLaws.laws(Function0.semigroup(Int.semigroup()), Function0 { 1 }, Function0 { 2 }, Function0 { 3 }, EQ),
+        MonoidLaws.laws(Function0.monoid(Int.monoid()), Function0 { 1 }, EQ),
```
Comment (pakoito) [inline modules/core/arrow-data/src/test/kotlin/arrow/data/Function0Test.kt]: Monoid implies semigroup, so there's no need to check both 🎉

File: modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/char.kt [hunk 2]
```diff
@@ -13,6 +11,10 @@ interface CharEqInstance : Eq<Char> {
   override fun Char.eqv(b: Char): Boolean = this == b
 }
 
+interface CharSemigroupInstance : Semigroup<Char> {
+  override fun Char.combine(b: Char): Char = this + b
```
Comment (pakoito) [inline modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/char.kt]: *squint* is that just...math?

File: modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/char.kt [hunk 3]
```diff
@@ -37,4 +39,4 @@ fun Char.Companion.eq(): Eq<Char> =
   object : CharEqInstance {}
 
 fun Char.Companion.order(): Order<Char> =
-  object : CharOrderInstance {}
\ No newline at end of file
+  object : CharOrderInstance {}
```
Comment (pakoito) [inline modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/char.kt]: nnnice

File: modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/const.kt [hunk 4]
```diff
@@ -91,4 +91,4 @@ class ConstContextPartiallyApplied<L>(val MA: Monoid<L>) {
 }
 
 fun <L> ForConst(MA: Monoid<L>): ConstContextPartiallyApplied<L> =
-  ConstContextPartiallyApplied(MA)
\ No newline at end of file
+  ConstContextPartiallyApplied(MA)
```
Comment (pakoito) [inline modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/const.kt]: yes

File: modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/function0.kt [hunk 5]
```diff
@@ -11,6 +11,25 @@ interface Function0FunctorInstance : Functor<ForFunction0> {
     fix().map(f)
 }
 
+@instance(Function0::class)
+interface Function0SemigroupInstance<O> : Semigroup<Function0<O>> {
+  fun SG(): Semigroup<O>
+
+  override fun Function0<O>.combine(b: Function0<O>): Function0<O> {
+    val f1 = this
```
Comment (pakoito) [inline modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/function0.kt]: you shouldn't need this. Use this@combine instead.

File: modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/char.kt (hunk 2 above)
Comment (bassjacob) [reply modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/char.kt]: nope, and now that I think about it, this instance is impossible

**Comment by bassjacob** [comment]
closing in favour of https://github.com/arrow-kt/arrow/pull/1219 which is easier than fixing conflicts

**Pull request closed without merging**